File: src/2025-02-23-deepclaude-try/python/exercises/practice/sgf-parsing/sgf_parsing.py
class SgfTree:
    def __init__(self, properties=None, children=None):
        self.properties = properties or {}
        self.children = children or []

    def __eq__(self, other):
        if not isinstance(other, SgfTree):
            return False
        for key, value in self.properties.items():
            if key not in other.properties:
                return False
            if other.properties[key] != value:
                return False
        for key in other.properties.keys():
            if key not in self.properties:
                return False
        if len(self.children) != len(other.children):
            return False
        for child, other_child in zip(self.children, other.children):
            if child != other_child:
                return False
        return True

    def __ne__(self, other):
        return not self == other


def parse(input_string):
    def parse_properties(content, index):
        """Parse SGF properties starting at index, return (properties, new_index)"""
        properties = {}
        while index < len(content):
            char = content[index]
            if char in ';()':
                break
            # Parse key
            if not char.isupper():
                if char == '[':
                    raise ValueError("properties without delimiter")
                raise ValueError("property must be in uppercase")
            
            key_end = index
            while key_end < len(content) and content[key_end].isupper():
                key_end += 1
            key = content[index:key_end]
            if not key.isupper():
                raise ValueError("property must be in uppercase")
            index = key_end
            
            # Parse values
            if index >= len(content) or content[index] != '[':
                raise ValueError("properties without delimiter")
            
            values = []
            while index < len(content) and content[index] == '[':
                index += 1
                value_end = index
                escaped = False
                while value_end < len(content):
                    if escaped:
                        escaped = False
                        value_end += 1
                    else:
                        if content[value_end] == '\\':
                            escaped = True
                            value_end += 1
                        elif content[value_end] == ']':
                            break
                        else:
                            value_end += 1
                    if value_end >= len(content):
                        raise ValueError("property without delimiter")
                
                # Process value according to SGF Text rules
                raw_value = content[index:value_end]
                processed = []
                i = 0
                while i < len(raw_value):
                    if raw_value[i] == '\\':
                        if i + 1 < len(raw_value):
                            next_char = raw_value[i + 1]
                            if next_char == '\n':
                                i += 2
                                continue
                            # Convert escaped whitespace to space
                            if next_char.isspace():
                                processed.append(' ')
                            else:
                                processed.append(next_char)
                            i += 2
                        else:
                            processed.append('\\')
                            i += 1
                    elif raw_value[i] == '\n':
                        processed.append('\n')
                        i += 1
                    elif raw_value[i].isspace():
                        processed.append(' ')
                        i += 1
                    else:
                        processed.append(raw_value[i])
                        i += 1
                
                values.append(''.join(processed))
                index = value_end + 1
            
            if not values:
                raise ValueError("property without delimiter")
            properties[key] = values
        
        return properties, index

    def parse_node(content, index):
        """Parse a single node starting at index, return (node, new_index)"""
        if index >= len(content) or content[index] != ';':
            raise ValueError("tree missing")
        
        index += 1  # Skip ';'
        properties, index = parse_properties(content, index)
        return SgfTree(properties, []), index

    def parse_tree(content, index):
        """Parse a tree or subtree starting at index, return (tree, new_index)"""
        if index >= len(content):
            raise ValueError("tree missing")
            
        # Parse root node
        root, index = parse_node(content, index)
        
        # Parse children
        children = []
        while index < len(content):
            char = content[index]
            if char == '(':
                # Parse variation
                depth = 1
                start = index
                index += 1
                while index < len(content) and depth > 0:
                    if content[index] == '(':
                        depth += 1
                    elif content[index] == ')':
                        depth -= 1
                    index += 1
                if depth > 0:
                    raise ValueError("tree missing")
                subtree = parse(content[start:index])
                children.append(subtree)
            elif char == ';':
                # Parse sequential node
                node, new_index = parse_node(content, index)
                if not children:
                    children.append(node)
                else:
                    children[-1].children.append(node)
                index = new_index
            elif char == ')':
                break
            else:
                raise ValueError("tree missing")
        
        root.children = children
        return root, index

    # Main parsing logic
    if not input_string:
        raise ValueError("tree missing")
    
    if not input_string.startswith('(') or not input_string.endswith(')'):
        raise ValueError("tree missing")
    
    content = input_string[1:-1]
    if not content:
        raise ValueError("tree with no nodes")
    
    tree, index = parse_tree(content, 0)
    if index < len(content):
        raise ValueError("tree missing")
    
    return tree

File: src/2025-02-23-deepclaude-try/java/exercises/practice/sgf-parsing/src/main/java/SgfParsing.java
import java.util.*;
import java.util.concurrent.atomic.AtomicInteger;

public class SgfParsing {
    public SgfNode parse(String input) throws SgfParsingException {
        if (input == null || input.isEmpty()) {
            throw new SgfParsingException("Empty input");
        }
        
        if (!input.startsWith("(") || !input.endsWith(")")) {
            throw new SgfParsingException("Tree missing outer parentheses");
        }
        
        // Remove outer parentheses
        String content = input.substring(1, input.length() - 1);
        if (content.isEmpty()) {
            throw new SgfParsingException("Empty tree");
        }
        
        AtomicInteger pos = new AtomicInteger(0);
        SgfNode root = parseNode(content, pos);
        
        // Check for unexpected content after parsing
        if (pos.get() < content.length()) {
            throw new SgfParsingException("Unexpected content after root node");
        }
        
        return root;
    }

    private SgfNode parseNode(String content, AtomicInteger pos) throws SgfParsingException {
        // Each node must start with semicolon
        if (pos.get() >= content.length() || content.charAt(pos.get()) != ';') {
            throw new SgfParsingException("Node must start with semicolon");
        }
        pos.incrementAndGet(); // Skip semicolon
        
        Map<String, List<String>> properties = new HashMap<>();
        
        // Parse properties until we hit a variation, child node, or end
        while (pos.get() < content.length()) {
            char c = content.charAt(pos.get());
            if (c == '(' || c == ';' || c == ')') break;
            
            // Parse property key (must be uppercase letters)
            if (!Character.isUpperCase(c)) {
                throw new SgfParsingException("Property must start with uppercase letter");
            }
            
            String key = parseKey(content, pos);
            List<String> values = parseValues(content, pos);
            
            if (values.isEmpty()) {
                throw new SgfParsingException("Property must have at least one value");
            }
            
            properties.put(key, values);
        }
        
        if (properties.isEmpty()) {
            throw new SgfParsingException("Node must have at least one property");
        }
        
        SgfNode node = new SgfNode(properties);
        
        // Parse variations (branches in parentheses)
        while (pos[0] < content.length() && content.charAt(pos[0]) == '(') {
            pos[0]++; // Skip opening parenthesis
            SgfNode variation = parseNode(content, pos);
            if (pos[0] >= content.length() || content.charAt(pos[0]) != ')') {
                throw new SgfParsingException("Unclosed variation");
            }
            pos[0]++; // Skip closing parenthesis
            node.addVariation(variation);
        }
        
        // Parse child node if present
        if (pos[0] < content.length() && content.charAt(pos[0]) == ';') {
            node.addChild(parseNode(content, pos));
        }
        
        return node;
    }
    
    private String parseKey(String content, AtomicInteger pos) {
        int start = pos.get();
        while (pos.get() < content.length() && Character.isUpperCase(content.charAt(pos.get()))) {
            pos.incrementAndGet();
        }
        return content.substring(start, pos.get());
    }
    
    private List<String> parseValues(String content, AtomicInteger pos) throws SgfParsingException {
        List<String> values = new ArrayList<>();
        
        while (pos.get() < content.length() && content.charAt(pos.get()) == '[') {
            pos.incrementAndGet(); // Skip opening bracket
            StringBuilder value = new StringBuilder();
            boolean escaped = false;
            
            while (pos.get() < content.length()) {
                char c = content.charAt(pos.get());
                
                if (escaped) {
                    if (c == '\n') {
                        // Skip escaped newline
                    } else if (Character.isWhitespace(c)) {
                        value.append(' ');
                    } else {
                        value.append(c);
                    }
                    escaped = false;
                } else if (c == '\\') {
                    escaped = true;
                } else if (c == ']') {
                    pos.incrementAndGet();
                    break;
                } else {
                    if (Character.isWhitespace(c) && c != '\n') {
                        value.append(' ');
                    } else {
                        value.append(c);
                    }
                }
                pos.incrementAndGet();
            }
            
            values.add(value.toString());
        }
        
        return values;
    }
}
